lib/elixir/test/elixir/kernel/expansion_test.exs

```diff
@@ -1653,6 +1657,21 @@ defmodule Kernel.ExpansionTest do
       end
     end
 
+    test "raises if the only clause other than do is else" do
+      assert_raise CompileError, ~r"\"else\" can't be used as the only clause", fn ->
+        code =
+          quote do
+            try do
+              :ok
+            else
+              other -> other
+            end
+          end
+
+        expand(code)
+      end
+    end
```

Comment:
We also need a test for this:

```elixir
def foo do
  :ok
else
  other -> other
end
```

and we need to make sure that we don't say "try" in the error message but rather definition or something similar. If you look at `elixir_def.erl`, you can see that we set the `{origin, _}` attribute in the `try` node, exactly so we know the origin of the try.

Reply:
I pushed an update, please have a look cause I'm not sure it's the best way of doing this. FWIW, there are other places where we say something bad happened in `try` but it really happened in `def`:

```
iex(2)> defmodule Foo do
...(2)> def foo() do
...(2)> :ok
...(2)> catch
...(2)> _ -> :catch
...(2)> rescue
...(2)> _ -> :rescue
...(2)> end
...(2)> end
warning: "catch" should always come after "rescue" in try
  iex:3
```

Should we fix them (separate PR)>

Reply:
> Should we fix them (separate PR)?

Up to you!

Reply:
Yep, separate PR! :)